nemo/utils/exp_manager.py

```diff
@@ -606,51 +609,76 @@ def check_resume(
     if resume_from_checkpoint:
         checkpoint = resume_from_checkpoint
     if resume_if_exists:
-        # Use <log_dir>/checkpoints/ unless `dirpath` is set
-        checkpoint_dir = Path(dirpath) if dirpath else Path(Path(log_dir) / "checkpoints")
-
-        # when using distributed checkpointing, checkpoint_dir is a directory of directories
-        # we check for this here
-        dist_checkpoints = [d for d in list(checkpoint_dir.glob("*")) if d.is_dir()]
-        end_dist_checkpoints = [d for d in dist_checkpoints if d.match("*end")]
-        last_dist_checkpoints = [d for d in dist_checkpoints if d.match("*last")]
-
-        end_checkpoints = end_dist_checkpoints if end_dist_checkpoints else list(checkpoint_dir.rglob("*end.ckpt"))
-        end_checkpoints = _filter_out_unfinished_checkpoints(end_checkpoints)
-        last_checkpoints = last_dist_checkpoints if last_dist_checkpoints else list(checkpoint_dir.rglob("*last.ckpt"))
-        last_checkpoints = _filter_out_unfinished_checkpoints(last_checkpoints)
-
-        if not checkpoint_dir.exists() or (not len(end_checkpoints) > 0 and not len(last_checkpoints) > 0):
-            if resume_ignore_no_checkpoint:
-                warn = f"There were no checkpoints found in checkpoint_dir or no checkpoint folder at checkpoint_dir :{checkpoint_dir}. "
-                if checkpoint is None:
-                    warn += "Training from scratch."
-                elif checkpoint == resume_from_checkpoint:
-                    warn += f"Training from {resume_from_checkpoint}."
-                logging.warning(warn)
+        '''
+        attach valid checkpoint path to trainer if current rank is rank zero of any data parallel groups
+        this limit to only global rank 0 process calling s3, instead of all processes calling s3
+        '''
+        local_rank = int(os.environ.get("LOCAL_RANK", 0))
```

Comment:
this should be more general if possible

Reply:
To clarify, this is referring to:
```
# If we are using S3 checkpointing, we want check_resume to only execute on a single rank to avoid throttling S3.

if rank == 0 or not S3Utils.is_s3_url(dirpath):
```

The reason why we have this is because at large-scale, S3 can get throttled due to every device attempting to list the files in the checkpoint directory, so we have only global rank 0 read, and broadcast to other ranks in the NeMoModelCheckpoint setup()
```
    def setup(self, trainer, pl_module, stage: str) -> None:
        if is_global_rank_zero():
            logging.debug("Removing unfinished checkpoints if any...")
            NeMoModelCheckpoint._remove_unfinished_checkpoints(self.dirpath)
        # Ensure that all ranks continue with unfinished checkpoints removed
        if torch.distributed.is_initialized():
            torch.distributed.barrier()
        super().setup(trainer, pl_module, stage)
        # When using S3 checkpointing, only Rank 0 has the checkpoint and model path set in exp_manager.
        # Sync the values across all ranks to ensure consistency.
        path = trainer.strategy.broadcast(trainer.ckpt_path)
        trainer.ckpt_path = path

        assert (
            trainer.checkpoint_callback == self
        ), f"This instance should be trainer.checkpoint_callback {trainer.checkpoint_callback} != {self}"
        self.last_model_path = trainer.strategy.broadcast(self.last_model_path)
``` 

Is the ask to make it more general to not do this, or is it to use the `is_global_rank_zero()` method instead?

Reply:
Updated to use NeMo's is_global_rank_zero instead

Reply:
thanks